statemachine/src/main/java/com/amazonaws/stepfunctions/cloudformation/statemachine/DefinitionProcessor.java

```diff
@@ -137,12 +147,12 @@ private static String convertDefinitionObjectToString(final Map<String, Object>
         }
     }
 
-    private static String transformDefinition(final String definitionString, final Map<String, String> resourceMappings) {
+    private static String transformDefinition(final String definitionString, final Map<String, Object> resourceMappings) {
```

Comment:
Opening the door to any Object type being accepted here might introduce some bugs.

Have we considered using a wrapper class for acceptance definition substitution types?

Reply:
If not, let's validate that a String, Double, or Boolean is provided. If any other unaccepted Object type is provided, we should throw an appropriate exception.

Reply:
This might be handled at the CloudFormation level. Did we test whether providing invalid types even make it to the handler? Or whether they fail immediately since they violate the schema?

Reply:
Doesn't look like we're testing that for substitutions yet.

Reply:
When providing invalid types, the resource handler would throw an error with a message something like `invalid State Machine Definition: 'SCHEMA_VALIDATION_FAILED: Expected value of type Integer at /TimeoutSeconds`
I think it is better to let customers input any object they want as long as the definition string (after substitution) can be parsed to valid json or yaml string. 
Similarly without using `DefinitionS3Location` and `DefinitionSubstitution`, customers are currently allowed to write definitions for their statemachine with invalid input types and try to create a statemachine (although it would result in a failure) 
What do you guys think about this?

Reply:
>I think it is better to let customers input any object they want as long as the definition string (after substitution) can be parsed to valid json or yaml string.

Why? The substitution object is converted to a [String](https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-stepfunctions/blob/42950ed4f0033a4a5eaf7943d1673697b340fdee/statemachine/src/main/java/com/amazonaws/stepfunctions/cloudformation/statemachine/DefinitionProcessor.java#L155) so it won't prevent the definition from being parsed to JSON or YAML. The error message above indicates it this type mismatch isn't caught at the CloudFormation level either.

At the very least, let's check that the objects are of type `String.class`, `Double.class`, or `Boolean.class` before passing the definition on.

Reply:
> Why? The substitution object is converted to a [String](https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-stepfunctions/blob/42950ed4f0033a4a5eaf7943d1673697b340fdee/statemachine/src/main/java/com/amazonaws/stepfunctions/cloudformation/statemachine/DefinitionProcessor.java#L155) so it won't prevent the definition from being parsed to JSON or YAML. The error message above indicates it this type mismatch isn't caught at the CloudFormation level either.

Since are three ways of providing the state machine definitions inside a CFN template, my initial thought was that all three methods should have consistent behaviour. 
But I agree with validating the object types, and I added them in my latest commit.